src/core/reference/include/openvino/reference/sign.hpp

```diff
@@ -17,9 +17,21 @@ constexpr T sign(const T v) {
     return static_cast<T>(static_cast<bool>(v));
 }
 
-template <class T, typename std::enable_if<ov::is_floating_point<T>() || std::is_signed<T>::value>::type* = nullptr>
+template <class T,
+          typename std::enable_if<std::is_floating_point<typename std::decay<T>::type>::value ||
+                                  std::is_signed<T>::value>::type* = nullptr>
 constexpr T sign(const T v) {
-    return static_cast<T>((T{0} < v) - (v < T{0}));
+    return static_cast<T>(std::isnan(static_cast<float>(v)) ? v : ((T{0} < v) - (v < T{0})));
```

Comment:
Why cast `v` to float?
I think this function should look like:

```suggestion
    return std::isnan(v) ? v : static_cast<T>((T{0} < v) - (v < T{0})));
```

Reply:
as I remember if none cast here, some of our windows CI will fail with the following error:
`[1636/8023] Building CXX object src\core\CMakeFiles\ngraph_obj.dir\Release\src\op\sign.cpp.obj
FAILED: src/core/CMakeFiles/ngraph_obj.dir/Release/src/op/sign.cpp.obj 
sccache C:\BuildTools\VC\Tools\MSVC\14.38.33130\bin\Hostx64\x64\cl.exe  /nologo /TP -DENABLE_PROFILING_ALL -DENABLE_PROFILING_FIRST_INFERENCE -DENABLE_PROFILING_ITT -DIE_BUILD_POSTFIX=\"\" -DIE_THREAD=IE_THREAD_TBB -DIMPLEMENT_OPENVINO_API -DIN_OV_COMPONENT -DIN_OV_CORE_LIBRARY -DOPENVINO_STATIC_LIBRARY -DOV_BUILD_POSTFIX=\"\" -DOV_NATIVE_PARENT_PROJECT_ROOT_DIR=\"openvino\" -DOV_THREAD=OV_THREAD_TBB -DSELECTIVE_BUILD_ANALYZER -DTBB_PREVIEW_WAITING_FOR_WORKERS=1 -DCMAKE_INTDIR=\"Release\" -IC:\actions-runner\_work\openvino\openvino\openvino\src\core\include -IC:\actions-runner\_work\openvino\openvino\openvino\src\core\src -IC:\actions-runner\_work\openvino\openvino\openvino\src\core\reference\include -IC:\actions-runner\_work\openvino\openvino\openvino\src\core\dev_api -IC:\actions-runner\_work\openvino\openvino\openvino\src\common\util\include -IC:\actions-runner\_work\openvino\openvino\openvino\thirdparty\pugixml\src -IC:\actions-runner\_work\openvino\openvino\openvino\src\core\shape_inference\include -IC:\actions-runner\_work\openvino\openvino\openvino\src\frontends\common\include -IC:\actions-runner\_work\openvino\openvino\openvino\src\common\transformations\include -IC:\actions-runner\_work\openvino\openvino\openvino\src\common\itt\include -IC:\actions-runner\_work\openvino\openvino\openvino\thirdparty\ittapi\ittapi\src\ittnotify -IC:\actions-runner\_work\openvino\openvino\openvino\thirdparty\ittapi\ittapi\include -IC:\actions-runner\_work\openvino\openvino\openvino\src\common\conditional_compilation\include /DWIN32 /D_WINDOWS /GR /EHsc /D_CRT_SECURE_NO_WARNINGS /D_SCL_SECURE_NO_WARNINGS /EHsc /Gy /W3 /bigobj /MP /wd4251 /wd4275 /O2 /Ob2 /DNDEBUG  /sdl /guard:cf -std:c++14 -MD -WX /showIncludes /Fosrc\core\CMakeFiles\ngraph_obj.dir\Release\src\op\sign.cpp.obj /FdC:\actions-runner\_work\openvino\openvino\openvino\bin\intel64\Release\ /FS -c C:\actions-runner\_work\openvino\openvino\openvino\src\core\src\op\sign.cpp
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt\corecrt_math.h(419): error C2668: 'fpclassify': ambiguous call to overloaded function
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt\corecrt_math.h(304): note: could be 'int fpclassify(long double) throw()'
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt\corecrt_math.h(299): note: or       'int fpclassify(double) throw()'
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt\corecrt_math.h(294): note: or       'int fpclassify(float) throw()'
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt\corecrt_math.h(419): note: while trying to match the argument list '(_Ty)'
        with
        [
            _Ty=int
        ]
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt\corecrt_math.h(419): note: the template instantiation context (the oldest one first) is
C:\actions-runner\_work\openvino\openvino\openvino\src\core\src\op\sign.cpp(46): note: see reference to function template instantiation 'bool ov::element::IfTypeOf<ov::element::Type_t::f16,ov::element::Type_t::f32,ov::element::Type_t::i32,ov::element::Type_t::i64,ov::element::Type_t::u32,ov::element::Type_t::u64>::apply<ov::op::sign::Evaluate,const _Ty&,_Ty&,size_t>(const std::string &,ov::element::Type_t,const _Ty &,_Ty &,size_t &&)' being compiled
        with
        [
            _Ty=ov::Tensor
        ]`